### PR TITLE
updates to book_changes RPC command

### LINKED ISSUE #5033: Invalid "success" response from book_changes (possible race condition)
## Issue Description
When querying for a recent ledger by index, sometimes the result doesn't match the expected format.

## Steps to Reproduce
Make WebSocket API requests such as the following, replacing the ledger index with the most recent one:

```json
{
  "id": "example_book_changes",
  "command": "book_changes",
  "ledger_index": 88376663
}
```

The problem is intermittent, so it may take some tries to reproduce. Be sure to keep updating the ledger index with the most recent one as new ledgers are closed.

## Expected Result
The API should either return a response in the expected format or an error message.

**Example valid response (with non-empty changes):**

```json
{
  "id": "example_book_changes",
  "result": {
    "changes": [
      {
        "close": "1303424095097822e-4",
        "currency_a": "XRP_drops",
        "currency_b": "rchGBxcD1A1C2tdxF6papQYZ8kjRKMYcL/BTC",
        "high": "1304070002477734e-4",
        "low": "1302405543037992e-4",
        "open": "1303084400776639e-4",
        "volume_a": "60000000",
        "volume_b": "0.000460371"
      }
    ],
    "ledger_hash": "45765B07A378A94FCB6CEDD3D498E0CC15C73468F5F15AAA779079E67A234EB9",
    "ledger_index": 88376645,
    "ledger_time": 770512422,
    "type": "bookChanges"
  },
  "status": "success",
  "type": "response"
}
```

**Example expected error response if the ledger index is too high:**

```json
{
  "error": "invalidParams",
  "error_code": 31,
  "error_message": "Ledger index too large",
  "id": "example_book_changes",
  "request": {
    "command": "book_changes",
    "id": "example_book_changes",
    "ledger_index": 88376663
  },
  "status": "error",
  "type": "response"
}
```

## Actual Result

In some cases the response claims to be a success but it contains the wrong data, apparently some sort of status update on the status of acquiring the requested ledger. Looks like a [Ledger Request Object](https://xrpl.org/docs/references/http-websocket-apis/admin-api-methods/logging-and-data-management-methods/ledger_request#ledger-request-object).

```json
{
  "id": "example_book_changes",
  "result": {
    "hash": "63C34CA1EE6CD80B89A035E077A0266BF2AF8F4A5CC729D1B595940D304BC05A",
    "have_header": true,
    "have_state": false,
    "have_transactions": false,
    "needed_state_hashes": [],
    "needed_transaction_hashes": [],
    "peers": 8,
    "timeouts": 1
  },
  "status": "success",
  "type": "response"
}
```

## Environment

Tested using the xrplcluster.com servers running 2.2.0-rc1, but I think this API hasn't been changed much if at all since #4212 two years ago so the problems probably exist in older versions.

The s1.ripple.com cluster mostly goes to Clio servers; I haven't been able to reproduce on there.

### PR DESCRIPTION
**Pull request by ckeshava**
## High Level Overview of Change

Updates to the `book_changes` RPC command.
Pending clarification, I believe this PR also fixes #5033  and #5036 (improves the latency of the command)

### Context of Change
1. accepts shortcut strings (`current, closed, validated`) as inputs to `ledger_index` parameter in the RPC command.
2. returns `validated: true/false` in the RPC output

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

### CODE REVIEW

File: src/xrpld/rpc/BookChanges.h
Comment (Bronek) [inline src/xrpld/rpc/BookChanges.h:1]: Can you pls do a drive-by fix and include the missing headers ?

```
#include <xrpl/json/json_value.h>
#include <xrpl/protocol/LedgerFormats.h>
#include <xrpl/protocol/STAmount.h>
#include <xrpl/protocol/STObject.h>
#include <xrpl/protocol/TxFormats.h>
#include <xrpl/protocol/jss.h>

#include <memory>
```

Comment (ckeshava) [reply src/xrpld/rpc/BookChanges.h:1]: I have specified these include's in a8a21bfe4

Did you obtain these `#include`'s based on any compiler's warning? Is this based on visual observation or any compiler special flag/setting?

Comment (ckeshava) [reply src/xrpld/rpc/BookChanges.h:1]: How do you use a unit test file "on its own"? I usually do `cmake --build .` inside the build directory. It automatically re-compiles the modified-source-code files.

Comment (Bronek) [reply src/xrpld/rpc/BookChanges.h:1]: A unit test for a header file on its own will have one `#include` , only for the header file being tested. Aside from that it can only `#include` files related to the unit test framework being used, but it must not include any other project files. If you have a test like that, and it fails to compile, then that's an indication that the file being tested is missing includes.

There are tools which do the equivalent testing for you, without actually creating any additional files in the project, for example `clangd` - that's how I find such errors.

Comment (ckeshava) [reply src/xrpld/rpc/BookChanges.h:1]: thanks for the explanation, that seems reasonable to me. I've removed some of the extraneous include's from BookChanges_test in a recent commit

**Comment by Bronek** [comment]
@godexsoft you might want to test this branch with clio.

**Review by Bronek — APPROVED** [review]
I love that it gets rid of `getLedgerByContext` call in a USER context, also the newly added unit tests seem to bump test coverage overall. Nice work !

**Comment by ckeshava** [comment]
@Bronek  thanks for the quick review!

What do you mean by USER context? I feel the `getLedgerByContext` should only be used by the consensus mechanism (and Full History Nodes) to collect+store all ledgers. Its in-efficient for a `tiny` or `medium` resources validator to engage in acquisition of ledgers.

The `getLedgerByContext` method can also be removed in the `ledger_request` RPC call. I expect to see similar issues in that RPC call. Check my comment here: https://github.com/XRPLF/rippled/issues/5036#issuecomment-2279269503

File: src/xrpld/rpc/handlers/BookOffers.cpp [hunk 1]
```diff
@@ -204,13 +204,13 @@ doBookOffers(RPC::JsonContext& context)
 Json::Value
 doBookChanges(RPC::JsonContext& context)
 {
-    auto res = RPC::getLedgerByContext(context);
+    std::shared_ptr<ReadView const> ledger;
 
-    if (std::holds_alternative<Json::Value>(res))
-        return std::get<Json::Value>(res);
+    Json::Value result = RPC::lookupLedger(ledger, context);
```
Comment (godexsoft) [inline src/xrpld/rpc/handlers/BookOffers.cpp:209]: 1. Does this change still yield the same error from checking ledger_index etc.?
2. Can we do the same in `LedgerRequest.cpp` and eliminate getLedgerByContext() entirely?

**Review by godexsoft — COMMENTED** [review]
I don't see anything that could impact Clio. 
My only concern is whether the new code still does the same error handling as the old did.

File: src/xrpld/rpc/handlers/BookOffers.cpp (hunk 1 above)
Comment (Bronek) [reply src/xrpld/rpc/handlers/BookOffers.cpp:209]: 1. @godexsoft the `ledger_index` handling (and errors) will be the same as in almost every other command, since the vast majority of commands use `lookupLedger`
2. No; the ADMIN method `ledger_request` which, with this PR, will be the only method using `getLedgerByContext()`, and it  needs to rely on this function in order to download ledgers from the other nodes.

**Review by godexsoft — APPROVED** [review]

**Review by Bronek — CHANGES_REQUESTED** [review]
@ckeshava please also update API-CHANGELOG.md for release 2.3 . You need to add section for release 2.3 and list the changes in behaviour of `book_changes` (basically making it like all other methods, but also will now return `ledgerNotFound` if the ledger is very old). I suggest asking @mDuo13 for help.

File: API-CHANGELOG.md [hunk 2]
```diff
@@ -28,6 +28,11 @@ The `network_id` field was added in the `server_info` response in version 1.5.0
 
 ## XRP Ledger server version 2.0.0
 
+### Updates in 2.3.0
+
+(Breaking Changes)
+- Changes to the `book_changes` RPC: This RPC now returns a `validated` field in its response, which was missing in prior versions. Additionally, `book_changes` RPC exhibits similar behavior as the other non-Admin RPC commands, insofar as ledger-retrieval is concerned. Specifically, if the requested ledger version is old, then a `ledgerNotFound` error is returned. This is a breaking change, because prior versions would attempt to acquire such an old ledger.
```
Comment (ckeshava) [inline API-CHANGELOG.md:34]: @mDuo13  @Bronek  If need be, please suggest modifications to this message.

**Review by Bronek — APPROVED** [review]

File: API-CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,11 @@ The `network_id` field was added in the `server_info` response in version 1.5.0
...
+
```
Comment (intelliot) [inline API-CHANGELOG.md:35]:
```suggestion
### Breaking change in 2.3

- `book_changes`: If the requested ledger version is not available on this node, a `ledgerNotFound` error is returned and the node does not attempt to acquire the ledger from the p2p network (as with other non-admin RPCs).

Admins can still attempt to retrieve old ledgers with the `ledger_request` RPC.

### Addition in 2.3

- `book_changes`: Returns a `validated` field in its response, which was missing in prior versions.

```

**Review by intelliot — APPROVED** [review]
Suggested a rewording to the changelog for consistency. I also think it's ok to merge as-is and improve the docs later on (which is easy to do).
